azure/src/main/java/org/apache/iceberg/azure/adlsv2/VendedAdlsCredentialProvider.java

```diff
@@ -0,0 +1,157 @@
+  private transient volatile Map<String, SimpleTokenCache> azureSasCredentialMap;
+  private transient volatile RESTClient client;
+
+  public VendedAdlsCredentialProvider(Map<String, String> properties) {
+    Preconditions.checkArgument(null != properties, "Invalid properties: null");
+    Preconditions.checkArgument(null != properties.get(URI), "Invalid URI: null");
+    this.properties = SerializableMap.copyOf(properties);
+    azureSasCredentialMap = Maps.newHashMap();
+  }
+
+  public String credentialForAccount(String storageAccount) {
+    Map<String, SimpleTokenCache> tokenCacheForAccountMap = azureSasCredentialMap();
+    if (tokenCacheForAccountMap.containsKey(storageAccount)) {
+      return tokenFromCache(tokenCacheForAccountMap.get(storageAccount));
+    } else {
+      SimpleTokenCache tokenCache =
+          new SimpleTokenCache(() -> Mono.fromSupplier(() -> sasTokenSupplier(storageAccount)));
+      tokenCacheForAccountMap.put(storageAccount, tokenCache);
+      return tokenFromCache(tokenCache);
+    }
+  }
+
+  private String tokenFromCache(SimpleTokenCache simpleTokenCache) {
+    return simpleTokenCache.getToken().map(AccessToken::getToken).block();
+  }
+
+  private AccessToken sasTokenSupplier(String storageAccount) {
+    LoadCredentialsResponse response = fetchCredentials();
+    List<Credential> adlsCredentials =
+        response.credentials().stream()
+            .filter(c -> c.prefix().contains(storageAccount))
+            .collect(Collectors.toList());
+    Preconditions.checkState(
+        !adlsCredentials.isEmpty(),
+        String.format("Invalid ADLS Credentials for storage-account %s: empty", storageAccount));
+    Preconditions.checkState(
+        adlsCredentials.size() == 1,
+        "Invalid ADLS Credentials: only one ADLS credential should exist per storage-account");
+
+    Credential adlsCredential = adlsCredentials.get(0);
+    checkCredential(adlsCredential, AzureProperties.ADLS_SAS_TOKEN_PREFIX + storageAccount);
+    checkCredential(
+        adlsCredential, AzureProperties.ADLS_SAS_TOKEN_EXPIRES_AT_MS_PREFIX + storageAccount);
+
+    String updatedSasToken =
+        adlsCredential.config().get(AzureProperties.ADLS_SAS_TOKEN_PREFIX + storageAccount);
+    Long tokenExpiresAtMillis =
+        Long.parseLong(
+            adlsCredential
+                .config()
+                .get(AzureProperties.ADLS_SAS_TOKEN_EXPIRES_AT_MS_PREFIX + storageAccount));
+
+    return new AccessToken(
+        updatedSasToken, Instant.ofEpochMilli(tokenExpiresAtMillis).atOffset(ZoneOffset.UTC));
+  }
+
+  private Map<String, SimpleTokenCache> azureSasCredentialMap() {
+    if (this.azureSasCredentialMap == null) {
+      synchronized (this) {
+        if (this.azureSasCredentialMap == null) {
+          this.azureSasCredentialMap = Maps.newHashMap();
+        }
+      }
+    }
+    return this.azureSasCredentialMap;
+  }
+
+  private RESTClient httpClient() {
```

Comment:
since the refreshing is handled now similarly to how it's done for GCS, I think we can simplify this to https://github.com/apache/iceberg/blob/9b1d18f72c185b6e53f90b778ca412960a64adf5/gcp/src/main/java/org/apache/iceberg/gcp/gcs/OAuth2RefreshCredentialsHandler.java#L102-L107. That way we don't have to keep `client` instance as a `transient volatile` field

Reply:
also we can wrap it's usage now in a try-with-resources block, similar to https://github.com/apache/iceberg/blob/9b1d18f72c185b6e53f90b778ca412960a64adf5/gcp/src/main/java/org/apache/iceberg/gcp/gcs/OAuth2RefreshCredentialsHandler.java#L60-L70